core/services/loaders/LoaderFactory.php

```diff
@@ -85,27 +86,37 @@ class LoaderFactory
     private static $loader;
 
 
-
     /**
-     * @param mixed $generator      provided during  very first instantiation in
-     *                              EE_Load_Espresso_Core::handle_request()
-     *                              otherwise can be left null
+     * @param EE_Registry|CoffeeShop   $generator   provided during very first instantiation in
+     *                                              BootstrapDependencyInjectionContainer::buildLoader()
+     *                                              otherwise can be left null
+     * @param ClassInterfaceCache|null $class_cache also provided during first instantiation
+     * @param ObjectIdentifier|null    $object_identifier
      * @return LoaderInterface
      * @throws InvalidArgumentException
-     * @throws InvalidInterfaceException
      * @throws InvalidDataTypeException
+     * @throws InvalidInterfaceException
      */
-    public static function getLoader($generator = null)
-    {
-        if (! LoaderFactory::$loader instanceof LoaderInterface) {
-            $generator = $generator !== null ? $generator : EE_Registry::instance();
+    public static function getLoader(
+        $generator = null,
+        ClassInterfaceCache $class_cache = null,
+        ObjectIdentifier $object_identifier = null
+    ) {
+        if (
+            ! LoaderFactory::$loader instanceof LoaderInterface
+            && ($generator instanceof EE_Registry || $generator instanceof CoffeeShop)
+            && $class_cache instanceof ClassInterfaceCache
+            && $object_identifier instanceof ObjectIdentifier
+        ) {
             $core_loader = new CoreLoader($generator);
             LoaderFactory::$loader = new Loader(
                 $core_loader,
                 new CachingLoader(
                     $core_loader,
-                    new LooseCollection('')
-                )
+                    new LooseCollection(''),
+                    $object_identifier
+                ),
+                $class_cache
             );
```

Comment:
Hmm, so it appears there's a change in logic here.  Prior to the changes if a generator wasn't provided and `$loader` was not an instance of `LoaderInterface` then the `$generator` would get set to `EE_Registry` and used for the new loader.  However, **now** it looks like _any_ of the _new_ incoming arguments to `getLoader` are null, there will be no loader instantiated (so if its already null, it will return null).  

It seems this has the potential to break things... or is that unlikely because core should always have instantiated a loader (even before any add-ons need it)?

Reply:
ya this happens during the very first step we take during bootstrapping at `plugins_loaded` priority 0.
Please note that it's also extremely detrimental to have more than one Loader floating around since that could result in duplicate objects that should be shared instances,
so having the construction of this object a bit complex is not a problem if that hinders the ability to create a second Loader.